src/main/kotlin/org/pkl/intellij/psi/PklModuleImpl.kt

```diff
@@ -237,31 +237,33 @@ class PklModuleImpl(viewProvider: FileViewProvider) :
       )
     }
 
-  /** Find the closest project to this module. */
-  override val pklProject: PklProject?
-    get() {
-      return cacheManager.getCachedValue(
+  override val isInPklProject: Boolean
+    get() = pklProjectDir != null
+
+  private val pklProjectDir
+    get(): VirtualFile? =
+      cacheManager.getCachedValue(
         this,
-        projectCacheKey,
+        projectDirCacheKey,
         {
-          val virtualFile = originalFile.virtualFile
-          if (virtualFile.fileSystem !is LocalFileSystem) {
-            return@getCachedValue null
-          }
-          var dir = virtualFile.parent
+          val dependency = project.pklProjectModificationTracker()
+          var dir = originalFile.parent
           while (dir != null) {
-            if (dir.findChild(PKL_PROJECT_FILENAME) != null) {
-              return@getCachedValue project.pklProjectService.getPklProject(dir)?.let {
-                CachedValueProvider.Result(it, project.pklProjectService)
-              }
+            if (dir.findFile(PKL_PROJECT_FILENAME) != null) {
+              return@getCachedValue CachedValueProvider.Result.createSingleDependency(
+                dir.virtualFile,
+                dependency
+              )
             }
             dir = dir.parent
           }
-          return@getCachedValue null
```

Comment:
This was the reason that project syncs were not resolving imports/types after the sync is complete.

The issue is that returning `null` back to `getCachedValue` will cached this null indefinitely and never be invalidated.

This introduces a new modification tracker just for PklProject files, and caches based on these PklProject files getting added/deleted/renamed/etc.